fix: prevent ThreadLocal memory leak in PhDefault, add cleanup and tests

karishka1222 opened the pull request:
#### Fix ThreadLocal Memory Leak in PhDefault
   
   **What was done:**
   - Fixed potential memory leak in `PhDefault` class related to `ThreadLocal NESTING` usage
   - Implemented proper cleanup mechanism using try-finally block in `take()` method
   - Added `cleanupNesting()` static method for manual ThreadLocal cleanup
   - Added comprehensive unit tests for multi-threaded scenarios
   - Updated TODO comment to reflect the fix
   
   **Why this is important:**
   - Addresses TODO #2251 which was discussing potential memory leaks
   - Prevents memory leaks in long-running threads and server applications
   - Improves thread safety and resource management
   - Follows object-oriented principles of proper resource cleanup
   
   **Testing:**
   - All existing tests pass
   - New tests cover normal nesting, exception handling, and multi-threading scenarios
   - Verified no memory leaks in concurrent usage patterns

0crat commented:
@karishka1222 Thank you for your contribution! 🙌 For [better Git practices](https://www.yegor256.com/2014/04/15/github-guidelines.html), we recommend naming branches after ticket numbers. Instead of "`fix/threadlocal-memory-leak-phdefault`", consider using "`4332`" for clarity and consistency. This aligns with our team's workflow and earns you points. Let's work together to improve our branch naming conventions!

<!-- It is [not a good idea](https://www.yegor256.com/2014/04/15/github-guidelines.html) to name Git branches the way you named this one: "`fix/threadlocal-memory-leak-phdefault`". You've earned -12 points. Next time, better give your branch the same name as the number of the ticket that you are solving. In this case, a perfect name, for example, would be "`4332`".  -->

karishka1222 commented:
@yegor256 Hello, can you check this pull request, please

karishka1222 commented:
@yegor256 Hello, I've fixed pipeline. Sorry for so many commits

h1alexbel commented on eo-runtime/src/main/java/org/eolang/PhDefault.java:
```diff
@@ -41,6 +41,10 @@ public class PhDefault implements Phi, Cloneable {
      *  variable can be removed, or, if this is not possible
      *  (see https://github.com/objectionary/eo/pull/1930), come up with another solution.
+     *
+     *  FIXED: ThreadLocal memory leak has been resolved by implementing proper cleanup
```
@karishka1222 I believe we can remove this note, since its redundant.

h1alexbel commented on eo-runtime/src/main/java/org/eolang/PhDefault.java:
```diff
@@ -41,6 +41,10 @@ public class PhDefault implements Phi, Cloneable {
      *  {@link PhDefault#NESTING} to prevent memory leaks. We should either find a place where this
```
@karishka1222 it seems that the PR resolves this puzzle, so we should remove it from the source code. Am I missing something?

h1alexbel commented on eo-runtime/src/main/java/org/eolang/PhDefault.java:
```diff
@@ -255,6 +267,15 @@ public final void add(final String name, final Phi attr) {
+     * This method should be called when a thread is about to terminate
+     * or when the ThreadLocal is no longer needed.
+     */
+    static void cleansupNesting() {
```
@karishka1222 looks like we introduced this method `cleansupNesting()` in order to be only used in tests. Since its redundant in our live code, let's move the method to the tests

h1alexbel commented on eo-runtime/src/test/java/org/eolang/PhDefaultTest.java:
```diff
@@ -17,7 +17,7 @@
  * @since 0.1
  */
-@SuppressWarnings("PMD.TooManyMethods")
+@SuppressWarnings({"PMD.TooManyMethods", "PMD.AvoidGodClass", "PMD.GodClass"})
```
@karishka1222 in PMD, it looks like that we don't have `PMD.AvoidGodClass` rule, so let's remove its suppression, since it has no effect

h1alexbel commented on eo-runtime/src/test/java/org/eolang/PhDefaultTest.java:
```diff
@@ -29,6 +29,11 @@ final class PhDefaultTest {
+    /**
+     * Name of context.
+     */
+    private static final String CONTEXT = "context";
```
@karishka1222 avoid static literals in the tests, instead - try to make all your test inputs - unique. Also read this: https://www.yegor256.com/2016/05/03/test-methods-must-share-nothing.html

h1alexbel reviewed (changes requested):
@karishka1222 check my comments, please

karishka1222 commented:
@h1alexbel I fixed the code based on your review.

h1alexbel commented on eo-runtime/src/test/java/org/eolang/PhDefaultTest.java:
```diff
@@ -483,6 +483,101 @@ void failsCorrectlyWhenTooManyAttributesPut() {
+    }
+
+    @Test
+    void verifiesCleanupNesting() {
```
@karishka1222 maybe we can rename the test to `worksAfterCleanup`? With the current name, it looks like we testing our `cleanUpNesting()` private method, specified in tests. WDYT?

h1alexbel commented on eo-runtime/src/test/java/org/eolang/PhDefaultTest.java:
```diff
@@ -539,6 +634,26 @@ private static class Int extends PhDefault {
                     )
                 )
             );
+            this.add(
```
@karishka1222 maybe we can create custom `Phi` instance for each test case instead of reusing `PhDefaultTest.Int`? Currently I see potential problems with sharing test state between multiple test cases

h1alexbel reviewed (changes requested):
@karishka1222 please check my comments

karishka1222 commented:
@h1alexbel I fixed it, I'm not quite sure that I fixed the creation of Phi instances correctly. If it's not the right solution, please describe in more detail what you meant. And one more thing, I added @since 0.1 to the 2 JavaDoc at the beginning of the PhDefaultTest class, because there was an error from qulice 
[INFO] Checkstyle: eo-runtime/src/test/java/org/eolang/PhDefaultTest.java[27]: Missing '@since' tag in class/interface comment (JavadocTagsCheck)

this is the only solution I've found for this. Although the strange thing is that there has never been this error before, and I have never changed the JavaDoc in this project. If I can fix the error in a more logical way, please tell me.

h1alexbel reviewed (approved):
@karishka1222 LGTM, thanks!

h1alexbel commented:
@yegor256 take a look, please

karishka1222 commented:
@yegor256 I fixed the conflict, but after that many pipelines failed, please tell me what the problem might be, although these pipelines never crashed

karishka1222 commented:
@yegor256 everything is fixed

yegor256 commented:
@karishka1222 thanks!
